exchange-v2/contracts/ExchangeBulkV2.sol

```diff
@@ -0,0 +1,215 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.7.6;
+pragma abicoder v2;
+
+import "./ExchangeV2Core.sol";
+import "./RaribleTransferManager.sol";
+import "@rarible/royalties/contracts/IRoyaltiesProvider.sol";
+import "@rarible/exchange-interfaces/contracts/IWyvernExchange.sol";
+
+contract ExchangeBulkV2 is ExchangeV2Core, RaribleTransferManager {
+    using SafeMathUpgradeable for uint;
+    using LibTransfer for address;
+
+    IWyvernExchange public wyvernExchange;
+
+    struct WyvernOrder {
+        address[7] _addrs;
+        uint[9]  _uints;
+        uint8[4]  _feeMethodsSidesKindsHowToCalls;
+        bytes  _calldataSell;
+        bytes  _replacementPatternSell;
+        bytes  _staticExtradataSell;
+        uint8[1]  _vs;
+        bytes32[2]  _rssMetadata;
+    }
+
+    struct WyvernOrders {
+        address[14] addrs;
+        uint[18]  uints;
+        uint8[8]  feeMethodsSidesKindsHowToCalls;
+        bytes  calldataBuy;
+        bytes  calldataSell;
+        bytes  replacementPatternBuy;
+        bytes  replacementPatternSell;
+        bytes  staticExtradataBuy;
+        bytes  staticExtradataSell;
+        uint8[2]  vs;
+        bytes32[5]  rssMetadata;
+    }
+
+    function __ExchangeBulkV2_init(
+        INftTransferProxy _transferProxy,
+        IERC20TransferProxy _erc20TransferProxy,
+        IWyvernExchange  _wyvernExchange,
+        uint newProtocolFee,
+        address newDefaultFeeReceiver,
+        IRoyaltiesProvider newRoyaltiesProvider
+    ) external initializer {
+        __Context_init_unchained();
+        __Ownable_init_unchained();
+        __TransferExecutor_init_unchained(_transferProxy, _erc20TransferProxy);
+        __RaribleTransferManager_init_unchained(newProtocolFee, newDefaultFeeReceiver, newRoyaltiesProvider);
+        __OrderValidator_init_unchained();
+        wyvernExchange = _wyvernExchange;
+    }
+
+    /*Transfer by ExchangeV2 array orders */
+    function matchOrdersBulk(
+        LibOrder.Order[] memory sellOrders,
+        bytes[] memory sellOrderSignatures,
+        address buyer
+    ) external payable {
+        uint totalAmount;
+        for (uint256 i = 0; i < sellOrders.length; i++) {
+            validateFull(sellOrders[i], sellOrderSignatures[i]);
+            LibOrder.Order memory buyerOrder;
+            buyerOrder.maker = buyer;
+            buyerOrder.makeAsset = sellOrders[i].takeAsset;
+            buyerOrder.takeAsset = sellOrders[i].makeAsset;
+            buyerOrder.dataType = 0xffffffff;
+            totalAmount += matchAndTransferBulk(sellOrders[i], buyerOrder);
+        }
+        if (msg.value > totalAmount) {
+            address(_msgSender()).transferEth(msg.value.sub(totalAmount));
```

Comment:
if returning change is a problem we can probably skip it

Reply:
I think it`s not a problem

Reply:
so return change works good